### PR TITLE
[SPARK-18394][SQL] Make an AttributeSet.toSeq output order consistent 

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr sorted output attributes on their name and exprId in `AttributeSet.toSeq` to make the order consistent.  If the order is different, spark possibly generates different code and then misses cache in `CodeGenerator`, e.g., `GenerateColumnAccessor` generates code depending on an input attribute order.

## How was this patch tested?
Added tests in `AttributeSetSuite` and manually checked if the cache worked well in the given query of the JIRA.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala [hunk 1]
```diff
@@ -121,7 +121,12 @@ class AttributeSet private (val baseSet: Set[AttributeEquals])
 
   // We must force toSeq to not be strict otherwise we end up with a [[Stream]] that captures all
   // sorts of things in its closure.
-  override def toSeq: Seq[Attribute] = baseSet.map(_.a).toArray.toSeq
+  override def toSeq: Seq[Attribute] = {
+    // We need to keep a deterministic output order for `baseSet` because this affects a variable
+    // order in generated code (e.g., `GenerateColumnAccessor`).
+    // See SPARK-18394 for details.
+    baseSet.map(_.a).toArray.sortBy { a => (a.name, a.exprId.id) }
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala]: If it needs to be a Seq, then should the `toArray` be `toSeq`? maybe I missed way it has to be an array first

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala]: Yea, as you suggested, I initially did so. But, I just kept [the original code](https://github.com/apache/spark/commit/c4787a3690a9ed3b8b2c6c294fc4a6915436b6f7#diff-75576f0ec7f9d8b5032000245217d233R105) cuz I was afraid this change wrongly affected the others. cc: @marmbrus

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala]: I though `map` should always return a strict collection. I think it is safe to sort immediately after that.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala]: ok, I fix in that way. Thanks!

**Review by hvanhovell — APPROVED** [review]

**Comment by viirya** [comment]
In title and description, `Attribute.toSeq` seems to be `AttributeSet.toSeq`?

**Comment by maropu** [comment]
oh..ya, my bad.... thanks.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AttributeSetSuite.scala [hunk 2]
```diff
@@ -78,4 +78,34 @@ class AttributeSetSuite extends SparkFunSuite {
     assert(aSet == aSet)
     assert(aSet == AttributeSet(aUpper :: Nil))
   }
+
+  test("SPARK-18394 keep a deterministic output order along with attribute names") {
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AttributeSetSuite.scala]: Modify this test case. Add a scenario in which the attribute set has two columns have the same name but different ids?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala [hunk 3]
```diff
@@ -162,7 +162,8 @@ class PruningSuite extends HiveComparisonTest with BeforeAndAfter {
       }.head
 
       assert(actualOutputColumns === expectedOutputColumns, "Output columns mismatch")
-      assert(actualScannedColumns === expectedScannedColumns, "Scanned columns mismatch")
+      assert(actualScannedColumns.sorted === expectedScannedColumns.sorted,
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: Could you add a comment to explain where we call `AttributeSet.toSeq`?

**Comment by gatorsmile** [comment]
LGTM except two minor comments.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AttributeSetSuite.scala (hunk 2 above)
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AttributeSetSuite.scala]: ok

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala [hunk 4]
```diff
@@ -162,7 +162,13 @@ class PruningSuite extends HiveComparisonTest with BeforeAndAfter {
       }.head
 
       assert(actualOutputColumns === expectedOutputColumns, "Output columns mismatch")
-      assert(actualScannedColumns === expectedScannedColumns, "Scanned columns mismatch")
+
+      // Scanned columns in `HiveTableScanExec` are generated by the `pruneFilterProject` method
+      // in `SparkPlanner` that internally uses `AttributeSet.toSeq`.
+      // Since we change an output order of `AttributeSet.toSeq` in SPARK-18394,
+      // we need to sort column names for a test below.
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: How about?

>  Scanned columns in `HiveTableScanExec` are generated by the `pruneFilterProject` method in `SparkPlanner`. This method internally uses `AttributeSet.toSeq`, in which the returned output columns are sorted by the names and expression ids.

Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: look good, I'll update soon.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by hvanhovell** [comment]
Merging to master. Thanks!

**Pull request closed without merging**